### ISSUE TITLE
gdal_translate resampling and complex data

### ISSUE THREAD

**Issue by piyushrpt** [OSGeo/gdal#791]
## Expected behavior and actual behavior.

gdal_translate with average option not working on complex datasets.

## Steps to reproduce the problem.

Here is a simple python test script to test the issue. I have tried the same with command line and that doesn't work either. All the imaginary parts of the output are zero. Other resampling options seem to work fine.

```python
from osgeo import gdal, gdal_array
import numpy as np
import matplotlib.pyplot as plt

#Synthetic input data
data = np.ones((300,300), dtype=np.complex64)
x = np.exp(1j * 2 * np.pi * np.arange(300)/60)
y = np.exp(1j * 2 * np.pi * np.arange(300)/75)
data[:,:] = x[:,None] * y[None,:]

#gdal_translate
srcds = gdal_array.OpenNumPyArray(data)
destds = gdal.Translate('', srcds, format='MEM', width=100, height=100, resampleAlg='average')
arr = destds.ReadAsArray()

srcds = None
destds = None

plt.figure()
plt.subplot(1,2,1)
plt.imshow(np.angle(data))
plt.subplot(1,2,2)
plt.imshow(np.angle(arr))
plt.show()
```

## Operating system

OS X 10.11.6

## GDAL version and provenance

GDAL 2.4.0dev, released 2018/99/99
via Macports

**Comment by piyushrpt** [comment]
An update - "mode" option also seems to have the same issue.

**Comment by piyushrpt** [comment]
I have the average and mode options working with changes to gdalwarpkernel.cpp, when all data is valid. 

However, getting the apps to recognize no data for complex values might need changes to VRTsources? Looks like gdal_translate_lib is setting up simple sources for operations and the nodata decision might be happening in RasterIOInternal?

**Comment by piyushrpt** [comment]
After more digging, I tracked down the issue to 

https://github.com/OSGeo/gdal/blob/e53f94dbbebc0f725280dcd2c696eb3db7856816/gdal/gcore/rasterio.cpp#L820-L822

https://github.com/OSGeo/gdal/blob/e53f94dbbebc0f725280dcd2c696eb3db7856816/gdal/gcore/rasterio.cpp#L918-L934

Looks like the nodata settings are completely ignored if the dataset is of type complex. Do you foresee any issues with rearranging this section of the code. Looks like **bUseWarp** essentially acts as **bIsComplex**, and maybe the whole section can be collapsed to use a common approach?

**Comment by rouault** [comment]
I guess this issue can be closed now, no ?

**Comment by piyushrpt** [comment]
Absolutely. We should have fixed this with #808 . Sorry for not following up on this.
